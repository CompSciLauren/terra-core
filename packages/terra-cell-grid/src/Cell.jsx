import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Cell.module.scss';
import CellGridUtils from './_CellGridUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Content to be displayed for the row cell
   */
  children: PropTypes.node,
  /**
   * Function callback for the ref of the td.
   */
  refCallback: PropTypes.func,
  /**
   * Function callback for the ref of the td.
   */
  width: PropTypes.shape({
    static: PropTypes.shape({
      value: PropTypes.number,
      unit: PropTypes.string,
    }),
    percentage: PropTypes.number,
    scalar: PropTypes.number,
  }),
};

const defaultProps = {
  children: [],
};

const Cell = ({
  children,
  refCallback,
  width,
  ...customProps
}) => {
  const cellClassNames = cx([
    'cell',
    customProps.className,
  ]);

  return (
    <div
      {...customProps}
      style={CellGridUtils.styleFromWidth(width)} // eslint-disable-line react/forbid-dom-props
      className={cellClassNames}
      ref={refCallback}
    >
      {children}
    </div>
  );
};

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default Cell;
