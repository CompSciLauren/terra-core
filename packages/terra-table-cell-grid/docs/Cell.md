# Cell

The Cell is a sub component to be used inside a cell grid.

## Usage

```jsx
import React from 'react';
import { Cell } from 'terra-table-cell-grid';

<Cell key="row-0-cell-0">
  Table row text
</Cell>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
