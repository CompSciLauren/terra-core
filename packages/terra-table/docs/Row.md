# Row

The Row is a sub component that groups table cells passed through children. Each row handles it's own selection state through props. The onSelect function that can be passed to the row will return the metaData associated to the cell grid. The metaData allows for greater control of state with the ability to associate any indicator or function with the row.

## Usage

```jsx
import React from 'react';
import { Row, Cell } from 'terra-table';

  <Row
    key="unique-react-key"
    isSelectable
    isSelected
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Cell>
      Cell Content
    </Cell>
    <Cell>
      Cell Content
    </Cell>
  </Row>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
