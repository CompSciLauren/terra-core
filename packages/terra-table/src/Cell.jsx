import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './Cell.module.scss';
import { styleFromWidth } from './TableUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Child content to be displayed for the row cell.
   */
  children: PropTypes.node,
  /**
   * Whether or not the cell should adhere to the tables paddingStyle.
   */
  isPadded: PropTypes.bool,
  /**
   * Function callback returning the html node for the cell.
   */
  refCallback: PropTypes.func,
  /**
   * Width of the cell. Should match header cell counter-part.
   */
  width: PropTypes.shape({
    /**
     * Static width that for the cell.
     */
    static: PropTypes.shape({
      /**
       * Numerical width value.
       */
      value: PropTypes.number.isRequired,
      /**
       * Valid css units are supported (i.e. 'px', 'rem', etc).
       */
      unit: PropTypes.string.isRequired,
    }),
    /**
     * Percentage width of the row for the cell.
     */
    percentage: PropTypes.number,
    /**
     * Relative scalar value of the cell's width compared to its sibling cells.
     */
    scalar: PropTypes.number,
  }),
};

const defaultProps = {
  children: [],
  isPadded: false,
};

const Cell = ({
  children,
  isPadded,
  refCallback,
  width,
  ...customProps
}) => {
  const cellClassNames = cx([
    'cell',
  ]);

  let content = children;
  if (isPadded) {
    content = (
      <div className={cx('container')}>
        {content}
      </div>
    );
  }

  return (
    <div
      {...customProps}
      style={styleFromWidth(width)} // eslint-disable-line react/forbid-dom-props
      className={customProps.className ? `${cellClassNames} ${customProps.className}` : cellClassNames}
      ref={refCallback}
      role="gridcell"
    >
      {content}
    </div>
  );
};

Cell.propTypes = propTypes;
Cell.defaultProps = defaultProps;

export default Cell;
