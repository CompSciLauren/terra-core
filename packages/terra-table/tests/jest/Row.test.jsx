import React from 'react';
import Row from '../../src/Row';
import Cell from '../../src/Cell';

// Snapshot test
it('should render a default row', () => {
  const shallowComponent = shallow(<Row><Cell /></Row>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a selectable row', () => {
  const shallowComponent = shallow(<Row isSelectable><Cell /></Row>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a selected row', () => {
  const shallowComponent = shallow(<Row isSelectable isSelected><Cell /></Row>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a striped row', () => {
  const shallowComponent = shallow(<Row isStriped><Cell /></Row>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a checkmark selectionStyle row', () => {
  const shallowComponent = shallow(<Row selectionStyle="checkmark"><Cell /></Row>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a chevron selectionStyle row', () => {
  const shallowComponent = shallow(<Row selectionStyle="chevron"><Cell /></Row>);
  expect(shallowComponent).toMatchSnapshot();
});

it('should render a row with onSelect', () => {
  const mockCallBack = jest.fn();

  const shallowComponent = shallow(
    <Row isSelectable onSelect={mockCallBack}><Cell /></Row>,
  );
  expect(shallowComponent).toMatchSnapshot();
  shallowComponent.find('.row').simulate('click');
  shallowComponent.find('.row').simulate('keydown', { nativeEvent: { keyCode: 13 } });
  shallowComponent.find('.row').simulate('keydown', { nativeEvent: { keyCode: 32 } });
  expect(mockCallBack.mock.calls.length).toEqual(3);
});

it('should render a row with refCallback', () => {
  const shallowComponent = shallow(
    <Row refCallback={jest.fn()}><Cell /></Row>,
  );
  expect(shallowComponent).toMatchSnapshot();
});
