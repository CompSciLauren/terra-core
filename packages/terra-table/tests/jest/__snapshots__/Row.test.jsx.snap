// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a checkmark selectionStyle row 1`] = `
<div
  className="divider-none row"
  role="row"
>
  <CheckMarkCell
    isDisabled={false}
    isPadded={true}
    isSelectable={false}
    isSelected={false}
  />
  <Cell />
</div>
`;

exports[`should render a chevron selectionStyle row 1`] = `
<div
  className="divider-none row"
  role="row"
>
  <Cell />
  <ChevronCell
    isPadded={true}
    isVisible={false}
  />
</div>
`;

exports[`should render a default row 1`] = `
<div
  className="divider-none row"
  role="row"
>
  <Cell />
</div>
`;

exports[`should render a row with onSelect 1`] = `
<div
  aria-selected={false}
  className="is-selectable divider-none row"
  data-row-show-focus="true"
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="row"
  tabIndex="0"
>
  <Cell />
</div>
`;

exports[`should render a row with refCallback 1`] = `
<div
  className="divider-none row"
  role="row"
>
  <Cell />
</div>
`;

exports[`should render a selectable row 1`] = `
<div
  aria-selected={false}
  className="is-selectable divider-none row"
  data-row-show-focus="true"
  onBlur={[Function]}
  onMouseDown={[Function]}
  role="row"
  tabIndex="0"
>
  <Cell />
</div>
`;

exports[`should render a selected row 1`] = `
<div
  aria-selected={true}
  className="is-selected is-selectable divider-none row"
  data-row-show-focus="true"
  onBlur={[Function]}
  onMouseDown={[Function]}
  role="row"
  tabIndex="0"
>
  <Cell />
</div>
`;

exports[`should render a striped row 1`] = `
<div
  className="is-striped divider-none row"
  role="row"
>
  <Cell />
</div>
`;
